issue #1, Seed identification API for Nachet frontend

JolanThomassin opened the pull request:
**Goal of the Pull Request:**
This pull request aims to introduce a function that **extracts the content from all the seeds** within the HTML content and converts them into a **usable JSON format**.

**Wanted result:**
```
[
    {
      "family": "Malvaceae",
      
      "common_name": "Velvetleaf",
      
      "duration_of_life_cycle": "Annual",
      
      "seed_or_fruit_type": "Seed",
      
      "identification_features":
       {
        "seed_length": "3.0 - 4.0 mm",
    
        "seed_width": "2.3 - 3.6 mm",
    
        "shape": "Heart-shaped seed with unequal lobes and a conspicuous notch at hilum end",
    
        "surface_texture": "Seed dull with minute reticulations and scattered tubercles,concentrated at the hilum",
    
        "colour": "Seed greyish brown with a lighter-coloured pubescence",
    
        "other_features": "Flap-like tissue covering the hilum may be present"
       },
       
       "images": [
        "https://inspection.canada.ca/DAM/DAM-plants-vegetaux/STAGING/images-images/invasive_plants_seed_factsheet_abutilon_theophrasti_01cnsh_1476382781946_eng.jpg",
        "https://inspection.canada.ca/DAM/DAM-plants-vegetaux/STAGING/images-images/invasive_plants_seed_factsheet_abutilon_theophrasti_01cnsh_1476382781946_eng.jpg" 
      ],
      
       "source": "https://inspection.canada.ca/plant-health/seeds/seed-testing-and-grading/seeds-identification/abutilon-theophrasti/eng/1472604373914/1472604374428"
    }
  ]
```
**Related Issue:**
- https://github.com/ai-cfia/louis-db/issues/1

[The website from which information is being crawled](https://inspection.canada.ca/plant-health/seeds/seed-testing-and-grading/seeds-identification/eng/1333136604307/1333136685768)

JolanThomassin commented:
**Missing data:**
- Similar species
- General Information
- Distribution

Currently, we have a function that searches for a seed **based on its common name**. We intend to modify this function to **search for seeds using their scientific names** (e.g., Abutilon theophrasti for Velvetleaf) and **retrieve information for all seeds**, not just one.

rngadam commented:
Start creating a Python test with relevant test cases based on the comments you've received and fixes you've done during development.

JolanThomassin commented:
**The objective here is to:**
- **Retrieve all HTML pages** about the seeds.
- Establish a **connection to the OpenAI Azure Chat**.
- **Request ChatGPT to parse the files** and provide **JSON output**.
- **Transform the JSON data** into a format suitable for the database.
- Establish **a normalized database structure** to store this data.

**Related commit:**
- Creation of the file ([9db48f9](https://github.com/ai-cfia/louis-db/pull/21/commits/9db48f9ee9e503114365cc876192cd49e9c76ad7))

JolanThomassin commented:
The code is making good progress. The next steps involve **adding the source and image information to the JSON**. However, a challenge I'm facing is that I don't have access to the source in the document I'm parsing, and it appears that the image links are local and therefore not useful.

Once I address this issue, my next task will be to parse all the seeds and store them in another location.

JolanThomassin commented:
_The text will undergo changes until the pull request is concluded._

**To summarize this Pull Request https://github.com/ai-cfia/louis-db/pull/21 :**

The objective was to obtain all the seeds from the [official government website](https://inspection.canada.ca/plant-health/seeds/seed-testing-and-grading/seeds-identification/eng/1333136604307/1333136685768) in a usable data format, specifically JSON. I explored various approaches, and while there is always room for code improvement, it appears that the main issue has been resolved. I successfully retrieved the information from the web and converted it into JSON format.

In the subsequent paragraph, I will provide a summary of the actions I have taken, explaining both their successes and limitations. It's important to note that the challenge I faced was primarily due to the website's quality. If the website had a properly formatted document, such as XML, certain aspects of the task would likely have been easier.

---
**Here are the approaches I attempted:**

1. My initial attempt was to accomplish everything solely using PostgreSQL. However, I encountered numerous challenges in terms of time, data quality, complexity, adaptability, and the difficulty of reusability and maintenance. This approach might have worked if the website had been perfectly structured, but there were too many exceptions and irregularities that made parsing the website data effectively impractical (75d8cd9).
2. I addressed the website's formatting issue by employing ChatGPT to generate JSON. Initially, I utilized SQL commands to retrieve a list of all seeds, then matched each seed with its respective page using the scientific name. After stripping away all HTML tags, I provided GPT with a JSON template containing the relevant text and requested it to populate it with the accurate data. My main challenge revolved around execution time and difficulty in obtaining image links. (9db48f9, aec1731, 059f0ae, 1ff7f2d).
3. I switched from ChatGPT version 4 to version 3.5 turbo. This adjustment resulted in a reduction of the time required per seed by a factor of 2 and also decreased the number of tokens used per seed. Remarkably, the accuracy of the results remained unchanged.
4. To incorporate images into the JSON, I initially attempted to achieve this solely using SQL and Python. However, the website's complexities made it exceedingly challenging to execute this task. Consequently, I introduced a section in which I sent the data to ChatGPT, along with images and their corresponding descriptions in the prompt. With this information, GPT was able to discern which images corresponded to the specific seed and which were related to similar species, thus effectively handling the image inclusion process. (e621c5e).
5. I modified the process of sending seeds by adopting an on-demand approach. Rather than retrieving information for all seeds from the beginning, we now fetch data only when necessary. This adjustment has significantly reduced execution time when parsing seeds that have already been seen. However, I encountered some issues during seed parsing, including instances where some seeds were missing and others were added when they shouldn't have been. (e621c5e).
6. I've updated the process as per your latest changes. Now, we're no longer searching for seed pages within the HTML content; instead, we're using the URLs. However, the challenge persists, as some URLs on the website are still not correctly formatted, preventing us from accessing the correct page for every seed. To address this issue, you've introduced a text file to manage problematic seeds separately, allowing for the implementation of an alternative method at a later stage. These adjustments should help improve the accuracy and effectiveness of the data retrieval process. (e621c5e).

---
**Statistics following the latest commit: (e621c5e)**
- Total seeds: 124
- Newly added seed: 2
- Skipped because already existing seed: 106
- Problematic seed found during search: 16

---
**Needed changes:**
 - I will alter the method for obtaining seed pages. We will retrieve them through URLs found on the main page where all the seeds are listed.

---
**Related Issue:**
- https://github.com/ai-cfia/louis-db/issues/1

melanie-fressard commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,150 @@
+    ) AS filtered_data
+    LIMIT {wanted_files_number};
+"""
+cursor.execute(query)
```
Je sais pas si c'est mieux mais tu pourrais mettre ta query directement dans le execute

JolanThomassin replied:
C'est compliqué, mais il vaut mieux les garder séparés. Cela peut aussi aider à éviter les attaques par injection SQL (dans le cas de formulaire web par exemple), même si ce n'est pas le cas ici.

melanie-fressard replied:
Oooh you're right

melanie-fressard commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,150 @@
+    
+    query = (
+    f"SELECT regexp_replace("
+    f"'{seed_full_url}', '.*seeds-identification/([^/]+).*', '\\1') AS seed_name;"
```
C'est pour quoi ça ?

JolanThomassin replied:
ette requête permet de récupérer le nom des graines en les extrayant des URL.

melanie-fressard replied:
Comme on en a parlé c'était le regex qui me posait problème mais c'est bon du coup

melanie-fressard commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,150 @@
+    4. Sends the query to the database and fetches the retrieved data.
+    5. Concatenates the cleaned content into a single 'page.'
+    6. Sends a request to the Azure OpenAI endpoint to get a response.
+    7. Processes the response, extracting the scientific name and saving it as a JSON file.
```
Attention t'as des warning de ruff, faut que tu les corrige

melanie-fressard commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,150 @@
+                image_descriptions[1] AS photo_description
+            FROM (
+                SELECT
+                    (regexp_matches(content, 'src="([^"]+)"', 'g')) AS image_links,
```
Je pense qut tu devrais peut etre commenter pourquoi tu fais des regex comme ça

JolanThomassin replied:
La fonction regex_matches en SQL permet d'extraire des morceaux de texte d'une colonne en fonction de règles définies par des expressions. En d'autres termes, elle te permet d'obtenir des parties spécifiques de la donnée qui correspondent à un modèle de texte que tu définis à l'aide de l'expression régulière. Par exemple, ici, on extrait toutes les images d'une page en utilisant regex_matches avec un modèle qui correspond aux balises des images.

melanie-fressard replied:
ok got it, merci :)

melanie-fressard commented on nachet-data/prompt/system_prompt.txt:
```diff
@@ -0,0 +1 @@
+Make sure to respect the format and the indentation. You are assisting in collecting information about a plant or organism in both English (EN) and French (FR). Respect the provided structure, leaving fields empty if necessary. For seed length and width, include floating-point numbers in an array. Even if it's a single word, place the duration_of_life_cycle in an array. Do not add, shorten or mod ...
```
Comment tu les utilise ? Je suis pas sure de bien comprendre le fonctionnement

JolanThomassin replied:
Dans la fonction principale de mon code, je récupère les éléments de saisie nécessaires, je les stocke sous forme de chaînes de caractères, et lorsque j'effectue ma requête via l'API, j'envoie ces éléments de saisie dans mon texte.

melanie-fressard commented on sql/2023-09-20-issue1-seed-identification-api-for-nachet-frontend.sql:
```diff
@@ -0,0 +1,167 @@
+                      POSITION('</p>' IN SUBSTRING(content, POSITION('<h2>Common Name</h2>' IN content) + LENGTH('<h2>Common Name</h2>'), 1000)) - 1),
+            '<[^>]+>',
+            '',
+            'g'
```
Tu peux m'expliquer ce que c'est ? J'ai du mal à comprendre

JolanThomassin replied:
Normalement, je devrais supprimer le code SQL que j'ai oublié.

melanie-fressard commented on sql/ToDelete.json:
```diff
@@ -0,0 +1,34 @@
+{
```
il s'appelle ToDelete, est ce que tu le laisse en exemple pour l'instant ?

JolanThomassin replied:
J'ai également oublié de le supprimer.

melanie-fressard reviewed (changes requested):
Good job, par contre il faudrait que tu corriges les warning venant de ruff tu peux just mettre des \ pour casser tes lignes trop longues

k-allagbe commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,151 @@
+                file_name = seed_name
+                file_name = nachet.decode_french_text(file_name)
+                file_name += ".json"
+# 
```
Unintended?

k-allagbe commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,151 @@
+                image_information += f"\nImage description: {image_descriptions}\n\n"
+
+            print("Sending request for summary to Azure OpenAI endpoint...\n")
+            response = openai.get_chat_answer(system_prompt, user_prompt, json_template, page, image_information)  # noqa: E501
```
Best fix the warning instead of suppressing it. If correctly configured, the Black extension fixes most of the lint issues on save (ctrl+s).

k-allagbe commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,151 @@
+
+import ailab.db as db
+import ailab.db.nachet as nachet
+from ailab.models import openai
```
Missing __init__.py in ailab.models folder?

k-allagbe commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,151 @@
+for url, seed_name in url_to_seed_mapping.items():
+    print(f"{seed_name}")
+
+def seed_identification_api(system_prompt, user_prompt, json_template):
```
Use verbs for function names to indicate actions in function names.

k-allagbe commented on ailab/db/nachet/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+    return content
+
+def json_file_exists(directory_path, file_name):
```
This looks more like a `utils` function than a nachet specific function. Also, the need for it is debatable.

k-allagbe commented on 2023-09-20-issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,151 @@
+            print("Sending request for summary to Azure OpenAI endpoint...\n")
+            response = openai.get_chat_answer(system_prompt, user_prompt, json_template, page, image_information)  # noqa: E501
+
+            # print("Chat answer: \n" + response.choices[0].message.content + "\n")
```
If not needed, best remove.

k-allagbe commented on 2023-10-06-all-json-dictionary:
```diff
@@ -0,0 +1,27 @@
+import os
```
This file is missing an extension.

rngadam commented on nachet-data/prompt/user_prompt.txt:
```diff
@@ -0,0 +1 @@
+Please complete the following template with accurate information about a plant or organism, the information are all in the followint text :
```
typo in following.

I would ask GPT to rewrite your prompts too ;-)

rngadam commented:
@JolanThomassin it fails linting with ruff. Please ensure it passes before reviews.

rngadam commented:
@JolanThomassin also please address reviewers previous comments by answering them ("done" or asking for clarification)

rngadam reviewed (changes requested):
please address other reviewers comments and linter problem first.

JolanThomassin commented:
@rngadam 

Lint should be working now. Some code had too many comments, but that's okay now. I have replied to or reacted to every comment, so I'm unsure what's wrong. I have added new comments to make sure that everything requested has been corrected or explained, so it should be fine now 👍

JolanThomassin commented:
With the last changes the code don't work anymore, I need to make some tests

rngadam commented on nachet-data/prompt/user_prompt.txt:
```diff
@@ -1 +1 @@
-Fill in this template with precise information about a plant or organism using the provided text.
\ No newline at end of file
+Fill in this template with precise information about a plant or organism using the provided text. And then translate it in the wanted language in the JSON (In our case, French as fr).
```
in the desired language

rngadam reviewed (changes requested):
prefixing files with YYYY-mm-dd is reserved for SQL files. Drop that for your .py file.

JolanThomassin commented:
I still see my comments as "Pending", I hope you can see them cause I can't find how to send them
![image](https://github.com/ai-cfia/louis-db/assets/98430140/9805efba-fa87-46c6-9d37-7a4f23a25615)

k-allagbe reviewed (commented):
Nice modularity. Only, I would go a bit further and separate database operations from your script logic.

JolanThomassin commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,181 @@
+
+def get_seed_list_url(cursor, wanted_files_number):
+    ### Get a list of all seeds URL
+    query = f"""
+        SELECT DISTINCT
+            (regexp_matches(filtered_content,
+            'href="([^"]*seeds-identification[^"]*)"', 'g'))[1] AS seeds_url
+        FROM (
+            SELECT content AS filtered_content
+            FROM html_content
+            WHERE md5hash = '1a365c39a8afd80d438ddf1bd3c9afed'
+        ) AS filtered_data
+        LIMIT {wanted_files_number};
+    """
```
You are right, even if we can't have SQL injection if a better way to do it. I changed it

JolanThomassin commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,181 @@
+            f"SELECT regexp_replace("
+            f"'{seed_full_url}', '.*seeds-identification/([^/]+).*', '\\1') AS sd_nme;"
+        )
+        cursor.execute(query)
```
Like discuss, I moved all queries into a file in nachet

k-allagbe commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,181 @@
+        print("\nCurrent seed : " + seed_name)
+        seed_json_path = seed_name + ".json"
+
+        file_path = os.path.join(SEED_DATA_PATH, seed_json_path)
```
Should be lower case inside the function.

JolanThomassin replied:
It is now

k-allagbe commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,181 @@
+        LIMIT {wanted_files_number};
+    """
+    cursor.execute(query)
+    list_seed_url = cursor.fetchall()
```
You didn't actually change the var name.

k-allagbe commented on ailab/db/nachet/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+    return content
+
+def decode_french_text(text):
+    return text.encode('latin1').decode('unicode-escape')
```
This function looks more like a utils or helper. I suggest using `text.encode('latin1').decode('unicode-escape')` directly in the caller.

JolanThomassin replied:
Done

k-allagbe commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,126 @@
+                file_name = nachet.decode_french_text(file_name)
+                file_name += ".json"
+
+            file_path = os.path.join(SEED_DATA_PATH, file_name)
+            with open(file_path, "w") as json_file:
```
What happens if data is not a dict?

JolanThomassin replied:
I add an else to this If

k-allagbe commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,126 @@
+
+            file_path = os.path.join(SEED_DATA_PATH, file_name)
+            with open(file_path, "w") as json_file:
+                json.dump(data, json_file, ensure_ascii=False, indent=4)
```
Before writing the json to file, I would suggest a verification. For instance with a function that checks if all the requiered fields from the template are indeed present in each item of the dict. What do you think?

JolanThomassin replied:
It might be challenging because the fields can sometimes be empty, and there are instances where we genuinely lack the necessary information. I've requested the API to allow for these blank entries. If something is missing inadvertently, it usually signifies an error, and typically, we would catch that before recording it. So I don't thing hadding a verification is a good options.

k-allagbe replied:
I see. However errors can still slip into the output. Like a typo in a field name. I guess it depends a lot on how tolerant the system that will use this function is to these errors. i suggest having a discussion about this with @rngadam. Marking this as resolved.

JolanThomassin commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,126 @@
+from ailab.db.nachet.seed_queries import query_get_webpage
+from ailab.db.nachet.seed_queries import query_get_images 
+
+website_url = "https://inspection.canada.ca"
```
I changed it

k-allagbe commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,126 @@
+    7. Processes the response, extracting the name and saving it as a JSON file.
+    """
+    for url, seed_name in url_to_seed_mapping.items():
+        print("\nCurrent seed : " + seed_name)
```
For instance, using `logging` allows you to control the level of output by setting the logging level, thus deactivating some outputs without commenting out the lines in the code.

JolanThomassin replied:
I switch to logging for production purposes, even though I'm not a fan of adding extra lines for something so straightforward.

k-allagbe replied:
The import was enough. The configuration was not necessary.

JolanThomassin commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,126 @@
+
+    nachet_db = db.connect_db()
+    with nachet_db.cursor() as cursor:
+        list_seed_url = seed_urls = query_seeds_urls(cursor, 1)
```
I forgot to remove the older one, now it' done

JolanThomassin reviewed (commented):
I'm not sure if you've noticed the mistake you pointed out in your previous review before, but please feel free to mention anything as they come up. You won't overwhelm me, so don't worry, and thank you for all your feedback.

k-allagbe commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,147 @@
+
+        print("\nList of selected seeds :")
+        for url, seed_name in url_to_seed_mapping.items():
+            print(f"{seed_name}")
```
Replace prints with logging.

JolanThomassin replied:
Done, all print are now logging

k-allagbe commented on issue1-seed-identification-api-for-nachet-frontend.py:
```diff
@@ -0,0 +1,147 @@
+
+    nachet_db = db.connect_db()
+    with nachet_db.cursor() as cursor:
+        seed_urls = query_seeds_urls(cursor, 3)
```
If seed_urls are not used anywhere else, they might as well be created inside the `create_seed_url_mapping` function.

JolanThomassin replied:
This is a separate query, and I believe it's a better approach to maintain each query as distinct entities.

k-allagbe reviewed (commented):
In my opinion the script is clean as is. Broader issues could be fixed in other GitHub issues.

rngadam reviewed (approved):
after f2f discussion LGTM. Missing @k-allagbe closing the open conversations after verifying fixes.

melanie-fressard commented on ailab/models/openai.py:
```diff
@@ -27,8 +27,8 @@ def safe_get(key):
-        input = tokens,
-        engine = OPENAI_API_ENGINE
+        input=tokens,
+        engine="ada"
```
OPENAI_API_ENGINE

JolanThomassin replied:
It's Done ! ^^

rngadam commented:
@JolanThomassin I suggest you resolve your own conversations you initiated in this PR, by commenting on what changed, or why you decided not to continue or creating follow up issues.
